src/gluonts/dataset/common.py

```diff
@@ -207,13 +213,144 @@ def files(self) -> List[Path]:
+                rec["source"] = SourceContext(
+                    source=rec[self._SOURCE_COL_NAME],
+                    row=rec[self._ROW_COL_NAME],
+                )
+                del rec[self._SOURCE_COL_NAME]
+                del rec[self._ROW_COL_NAME]
+            else:
+                rec["source"] = SourceContext(
+                    source="arrow.Table", row=row_number
+                )
+            rec = self._process(rec)
+            yield rec
+            row_number += 1
+
+    def __len__(self):
+        return len(self.table)
+
+    @classmethod
+    def write_table_from_records(
+        cls,
+        it: Iterable[DataEntry],
+        file_path: Union[str, Path],
+        chunk_size=100,
+    ) -> None:
+        """
+        Write time series records as an arrow dataset to the file_path.
+        """
+        file_path = Path(file_path)
+        msg = f"Converting to arrow dataset: {file_path}"
+        l: Optional[int] = None
+        try:
+            l = len(it)  # type: ignore
+        except:
+            l = None
+        with ArrowWriter(file_path, chunk_size=chunk_size) as writer:
+            for rec in tqdm(it, total=l, desc=msg):
+                rec = rec.copy()
+                float_np_fields = [
+                    "target",
+                    "feat_static_real",
+                    "feat_dynamic_real",
+                ]
+                for field in float_np_fields:
+                    if field in rec:
+                        rec[field] = np.asarray(rec[field], dtype=np.float32)
+                int_np_fields = ["feat_static_cat", "feat_dynamic_cat"]
+                for field in int_np_fields:
+                    if field in rec:
+                        rec[field] = np.asarray(rec[field], dtype=np.int32)
+                if "start" in rec:
+                    rec["start"] = pd.Timestamp(rec["start"])
+                if "source" in rec:
+                    del rec["source"]
+                writer.write_record(rec)
+
+
+class MemmapArrowDataset(ArrowDataset):
+    def __init__(self, files: List[Path], freq: str, chunk_size: int):
+        """
+        Arrow dataset using memory mapped files that closes the files when the object is deleted.
+        """
+        self.files = files
+        self.mmaps = [pa.memory_map(str(p)) for p in files]
+        tables = []
+        for file_path, mm in zip(files, self.mmaps):
+            t = pa.ipc.open_stream(mm).read_all()
+            source_col = pa.repeat(str(file_path), len(t)).dictionary_encode()
+            t = t.append_column(self._SOURCE_COL_NAME, source_col)
+            row_col = pa.array(np.arange(len(t)))
+            t = t.append_column(self._ROW_COL_NAME, row_col)
+            tables.append(t)
+        if len(tables) > 1:
+            table = pa.concat_tables(tables)
+        else:
+            table = tables[0]
+        super().__init__(table, freq=freq, chunk_size=chunk_size)
+
+    def __del__(self):
```

Comment:
I don't like relying on `__del__`.

But can we at least have a `close()` method which is then called in `__del__`?

Reply:
Yes, will do. I only added this cleanup because something there is an issue on windows, where you cannot delete a directory with a memmaped file.